spyderlib/plugins/shortcuts.py

```diff
@@ -7,179 +7,505 @@
+                self.npressed -= 1
+            debug_print('decrementing')
+        if modifiers & Qt.AltModifier:
+            key += Qt.ALT
+        if modifiers & Qt.MetaModifier:
+            key += Qt.META
+
+        self.keys.add(key)
+
+    def toggle_state(self):
+        """Switch between shortcut entry and Accept/Cancel shortcut mode."""
+        self.edit_state = not self.edit_state
+
+        if not self.edit_state:
+            self.text_new_sequence.setEnabled(False)
+            if self.button_ok.isEnabled():
+                self.button_ok.setFocus()
+            else:
+                self.button_cancel.setFocus()
+        else:
+            self.text_new_sequence.setEnabled(True)
+            self.text_new_sequence.setFocus()
+
+    def nonedit_keyrelease(self, e):
+        """Key release event for non-edit state."""
+        key = e.key()
+
+        if key == Qt.Key_Escape:
+#            self.close()
+            pass
+
+        if key in [Qt.Key_Left, Qt.Key_Right, Qt.Key_Up,
+                   Qt.Key_Down]:
+
+            if self.button_ok.hasFocus():
+                self.button_cancel.setFocus()
+            else:
+                self.button_ok.setFocus()
+
+    def keyReleaseEvent(self, e):
+        """Qt override"""
+        self.npressed -= 1
+        if self.npressed <= 0:
+            if len(self.keys) == 1 and list(self.keys)[0] == Qt.Key_Tab:
+                self.toggle_state()
+
+            if not self.edit_state:
+                self.nonedit_keyrelease(e)
+            else:
+                debug_print('keys: {}'.format(self.keys))
+                if self.keys:
+                    self.validate_sequence()
+                self.keys = set()
+                self.key_modifiers = set()
+                self.key_non_modifiers = list()
+                self.key_text = list()
+                self.npressed = 0
+
+    def check_conflicts(self):
+        """Check shortcuts for conflicts"""
+        conflicts = []
+        for index, shortcut in enumerate(self.shortcuts):
+            sequence = str(shortcut.key)
+            if sequence == self.new_sequence and \
+                (shortcut.context == self.context or shortcut.context == '_' or
+                 self.context == '_'):
+                conflicts.append(shortcut)
+        return conflicts
+
+    def update_warning(self, reset=False):
+        """ """
+        conflicts = self.check_conflicts()
+        if reset:
+            conflicts = []
+
+        widget = self.helper_button
+
+        if conflicts:
+            tip_title = _('The new entered shorcut conflicts with:') + '\n'
+            tip_body = ''
+            for s in conflicts:
+                tip_body += ' - {0}: {1}\n'.format(s.context, s.name)
+            tip = '{0}{1}'.format(tip_title, tip_body)
+            widget.setIcon(get_std_icon('MessageBoxWarning'))
+            widget.setToolTip(tip)
+            QToolTip.showText(widget.mapToGlobal(QPoint(0, 5)), tip)

```

Comment:
@blink1073 can you take care of this?, it is very annoying to make a PR against your PR to fix all the "details"...